src/Passes/Source/AllocationManager/IAllocationManager.hpp

```diff
@@ -0,0 +1,74 @@
+#pragma once
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "Llvm/Llvm.hpp"
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace microsoft
+{
+namespace quantum
+{
+    /// Interface class for allocation management. This interface provides means to allocate and release
+    /// statically allocated resources such as qubits and results. In a future version, it may be
+    /// extended with get and store in order to support Arrays and Tuples.
+    class IAllocationManager
+    {
+      public:
+        using Address              = uint64_t;    ///< Value type for address
+        using Index                = uint64_t;    ///< Index type used to access an array element.
+        using String               = std::string; ///< String type used for naming segments.
+        using AllocationManagerPtr = std::shared_ptr<IAllocationManager>; ///< Pointer interface.
+
+        // Construction, moves and copies
+        //
+        IAllocationManager(IAllocationManager const&) = delete;
+        IAllocationManager(IAllocationManager&&)      = delete;
+        IAllocationManager& operator=(IAllocationManager const&) = delete;
+        IAllocationManager& operator=(IAllocationManager&&) = delete;
+
+        virtual ~IAllocationManager();
+
+        // Interface
+        //
+
+        /// Abstract member function to allocate an element or sequence of elements. The developer
+        /// should not assume continuity of the address segment as this is not guaranteed. Note this
+        /// function may throw if allocation is not possible.
+        virtual Address allocate(String const& name = "", Index const& size = 1) = 0;
```

Comment:
FYI: 
I feel that the second parameter specifies the _number of elements_ in the sequence to be allocated (rather than _size in bytes_). 

For my own code I prefer to make a clear distinction between 
the size in bytes (for which I use the word "size")
and the number of elements in a container/array (for which I use either the word "length" or the word "count" or any word other than "size").

Some more details are [here](https://github.com/kuzminrobin/code_review_notes/blob/master/cpp_design_bookmarks.md#distinguish-between-size-and-length).

Reply:
Yes, this is a good point. It would be more accurate to call it `count`. In general, there is still some questions as to where the allocation manager should be used: Right now it is only used for Qubits and Results. If we extend this to also take care of Arrays then the argument should be the byte size and otherwise, it suffices to have the number of elements.

I've refactored to `count`.